Handle `mlflow.chat.message` deprecation in frontend logic

frontsideair opened the pull request:
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/frontsideair/mlflow/pull/16760?quickstart=1)

#### Install mlflow from this PR

```
# mlflow
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/16760/merge
# mlflow-skinny
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/16760/merge#subdirectory=libs/skinny
```

For Databricks, use the following command:

```
%sh curl -LsSf https://raw.githubusercontent.com/mlflow/mlflow/HEAD/dev/install-skinny.sh | sh -s pull/16760/merge
```

</p>
</details>

### Related Issues/PRs

<!-- Uncomment 'Resolve' if this PR can close the linked items. -->
Related to #16762

### What changes are proposed in this pull request?

Implement logic to parse chat messages in the frontend, preparing for the deprecation of `mlflow.chat.message` field.

### How is this PR tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

<!-- Attach code, screenshot, video used for manual testing here. -->

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Release Notes

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

<!-- Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change. -->

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/prompt`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

<!-- patch -->

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

frontsideair committed:
Handle OpenAI Responses API inputs and outputs for chat messages

Signed-off-by: Fatih Altinok <fatihaltinok@live.com>

B-Step62 commented on mlflow/server/js/src/shared/web-shared/model-trace-explorer/ModelTraceExplorer.utils.tsx:
```diff
@@ -1011,6 +1004,11 @@ export const normalizeConversation = (input: any): ModelTraceChatMessage[] | nul
       return openAIResponsesOutput;
     }
 
+    const openAIResponsesInput = normalizeOpenAIResponsesInput(input);
```
@daniellok-db @frontsideair In the Python SDK side, I'm adding a new attribute `mlflow.message.format` indicates what library created the LLM span e.g., `langchain`. Thoughts on refactoring this logic to check that attribute and switch to respective parsing logic?
Since we will add 5 more different parsing logic here, checking sequentially seems to be inefficient and might have false positive.

frontsideair replied:
I'd be happy to support this pattern, I agree about sequentially trying everything not being great.

daniellok-db replied:
Oh yes that would be wonderful. We can fall back to sequential parsing for backward compatibility if the attribute is not present

frontsideair committed:
Support DSPy chat inputs/outputs

Signed-off-by: Fatih Altinok <fatihaltinok@live.com>

frontsideair committed:
Handle gemini chat inputs and outputs

Signed-off-by: Fatih Altinok <fatihaltinok@live.com>

frontsideair committed:
Handle Anthropic chat inputs and outputs

Signed-off-by: Fatih Altinok <fatihaltinok@live.com>

frontsideair committed:
Refactor chat normalizes to their own directories

Signed-off-by: Fatih Altinok <fatihaltinok@live.com>

frontsideair committed:
Refactor the code to use lodash more

Signed-off-by: Fatih Altinok <fatihaltinok@live.com>

frontsideair committed:
Move normalizers to chat-utils

Signed-off-by: Fatih Altinok <fatihaltinok@live.com>

daniellok-db reviewed (approved):
Thanks so much for the reorganization! It's something I wanted to do for a long time as well, the main utils file was getting way too bloated. 

The code LGTM, did you manage to get the data you needed to test all the cases? If so it'd be great to just have a simple test for each flavor that runs the normalizer functions on a sample input (can just copy paste the real JSON from test API request/responses) and asserts that a chat message is generated.

I'll try to pull this PR to test locally myself as well, but assuming the logic works let's merge!

daniellok-db reviewed (commented):
re: the test failures, it looks like the current [mock openai output](https://github.com/mlflow/mlflow/blob/0aa49a83cd5139ee04ad29392fef07545a0eafe4/mlflow/server/js/src/shared/web-shared/model-trace-explorer/chat-utils/openai.test-utils.ts#L39) is literally the contents of the "outputs" field, which is why the test is failing

this is a quirk of how the summary tab is rendered: each field in the input is parsed separately, whereas chat tab data processes the whole input at once and just extracts relevant fields.

i don't think it blocks this PR but we should think if there's an easy way to make this logic work for summary view easily.

daniellok-db commented:
we don't have to merge now! i think it's good to finish up the improvements / test fixes etc. i think @B-Step62's PR hasn't merged yet, we can follow up on that separately (or we can take it over if it doesn't make it in by today)

frontsideair committed:
Add support to display image responses in OpenAI Responses API

Signed-off-by: Fatih Altinok <fatihaltinok@live.com>

frontsideair committed:
Fix OpenAI Responses API image input handling

Signed-off-by: Fatih Altinok <fatihaltinok@live.com>

frontsideair committed:
Fix broken tests

Deleted OpenAI streaming output tests because the mock was wrong and I
didn't have any replacement immediately

Signed-off-by: Fatih Altinok <fatihaltinok@live.com>

frontsideair committed:
Add an anthropic input/output chat trace test

Signed-off-by: Fatih Altinok <fatihaltinok@live.com>

frontsideair committed:
Add gemini input/output test for trace chats

Signed-off-by: Fatih Altinok <fatihaltinok@live.com>

frontsideair committed:
Add tests for the rest of the chat tools

Signed-off-by: Fatih Altinok <fatihaltinok@live.com>

frontsideair commented:
| Provider | Screenshot |
|--------|--------|
| OpenAI (responses) | <img width="595" height="705" alt="Screenshot 2025-07-18 at 18 58 29" src="https://github.com/user-attachments/assets/11e8be7e-9c07-4208-9074-27fba70d6e6f" /> |
| OpenAI (responses) | <img width="595" height="705" alt="Screenshot 2025-07-18 at 18 58 39" src="https://github.com/user-attachments/assets/ec7f6ad6-cf4f-4858-ba11-42f405e72b79" /> |
| Anthropic | <img width="595" height="705" alt="Screenshot 2025-07-18 at 18 58 54" src="https://github.com/user-attachments/assets/a8d5d627-bc57-447e-aab2-91c1c79c6329" /> |
| Gemini | <img width="595" height="705" alt="Screenshot 2025-07-18 at 18 59 31" src="https://github.com/user-attachments/assets/2414be9c-2fcc-4a59-8172-7099eb349572" /> |
| DSPy | <img width="595" height="705" alt="Screenshot 2025-07-18 at 18 59 47" src="https://github.com/user-attachments/assets/48391bc0-b253-4980-9953-ba85672fa11a" /> | 
| LangChain | <img width="595" height="705" alt="Screenshot 2025-07-18 at 18 59 58" src="https://github.com/user-attachments/assets/549c1747-899a-4cc2-a4a3-a241def2dee6" /> | 
| LlamaIndex | <img width="595" height="705" alt="Screenshot 2025-07-18 at 19 00 13" src="https://github.com/user-attachments/assets/c72d0be5-4ee0-4c3a-ba65-f10f3b2d9ac3" /> |
